<html>
   <head>
      <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
      <link href="style.css" rel="stylesheet" type="text/css">
      <title>SchemaDoc:
         Gene
      </title>
   </head>
   <body><a href="index.html">Table Index</a><h1><a name="Gene"></a>Gene
      </h1>
      <hr>Defines a notion of gene that is limited to procaryotic aspects of<br>genes. Later versions of the warehouse will expand this definition to include<br>eukaryotic aspects of genes. Separate tables define associations between a gene and<br>(1) its biological source(s),<br>(2) its protein product(s), if any, and<br>(3) its RNA product(s), if any.<br><h2>Columns</h2>
      <table border="1">
         <tr>
            <th>Column</th>
            <th>MySQL Type</th>
            <th>Oracle Type</th>
            <th>Nullable</th>
            <th>Description</th>
         </tr>
         <tr>
            <td>WID</td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>No</td>
            <td>Warehouse identifier this gene.<br></td>
         </tr>
         <tr>
            <td>Name</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(255)</td>
            <td>Yes</td>
            <td>Name of the gene.<br></td>
         </tr>
         <tr>
            <td><a href="NucleicAcid.html">NucleicAcidWID</a></td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>Reference to the NucleicAcid molecule (replicon) this gene resides upon.<br></td>
         </tr>
         <tr>
            <td>SubsequenceWID</td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>Reference to the Subsequence containing the nucleotide sequence of this Gene.<br></td>
         </tr>
         <tr>
            <td>Type</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(100)</td>
            <td>Yes</td>
            <td>Describes the type of molecule which is known to be *ultimately* produced by this gene<br>enumerated values (polypeptide, pre-mRNA, rRNA, tRNA, etc)<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>unknown</b> -
                  used when transcriptional status of gene is unknown<br><b>pre-RNA</b> -
                  used when there is no evidence that a mature RNA is ultimately produced by the gene<br><b>mRNA</b> -
                  used when there is evidence that a mature mRNA is ultimately produced by the gene<br><b>rRNA</b> -
                  used when there is evidence that a mature rRNA is ultimately produced by the gene<br><b>tRNA</b> -
                  used when there is evidence that a mature tRNA is ultimately produced by the gene<br><b>snRNA</b> -
                  used when there is evidence that an snRNA is ultimately produced by the gene<br><b>scRNA</b> -
                  ????<br><b>polypeptide</b> -
                  used when there is evidence that the ultimate product of this gene is proteinaceous<br><b>snoRNA</b> -
                  ???<br><b>other</b> -
                  catchall ?<br></div>
            </td>
         </tr>
         <tr>
            <td>GenomeID</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(35)</td>
            <td>Yes</td>
            <td>Unique ID assigned to this gene, such as by a genome project<br></td>
         </tr>
         <tr>
            <td>CodingRegionStart</td>
            <td>INT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>Base position of start of coding region. Start is always less than End,<br>except for genes that wrap around the origin of a circular chromosome.<br></td>
         </tr>
         <tr>
            <td>CodingRegionEnd</td>
            <td>INT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>Base position of end of coding region.<br>Indexes the stop codon that terminates the gene.<br></td>
         </tr>
         <tr>
            <td>CodingRegionStartApproximate</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(10)</td>
            <td>Yes</td>
            <td>Indicates that the Start position of the coding region is an approximate value.<br>It could be 'gt' for greater than, 'lt' for less than and 'ne' to indicate that it is not<br>equal. This is a controlled vocabulary.<br></td>
         </tr>
         <tr>
            <td>CodingRegionEndApproximate</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(10)</td>
            <td>Yes</td>
            <td>Indicates that the End position of the coding region is an approximate value.<br>It could be 'gt' for greater than, 'lt' for less than and 'ne' to indicate that it is not<br>equal. This is a controlled vocabulary.<br></td>
         </tr>
         <tr>
            <td>Direction</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(25)</td>
            <td>Yes</td>
            <td>Direction of transcription as defined in the enumeration table<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>unknown</b> -
                  unknown the strand being transcribed is unknown<br><b>forward</b> -
                  the plus strand is being transcribed<br><b>reverse</b> -
                  the minus strand is being transcribed<br><b>forward_and_reverse</b> -
                  both strands are being transcribed<br><b>undefined_value</b> -
                  UNDEFINED; the NCBI documentation does not define this value<br></div>
            </td>
         </tr>
         <tr>
            <td>Interrupted</td>
            <td>CHAR(1)</td>
            <td>CHAR(1)</td>
            <td>Yes</td>
            <td>'T' if the gene is interrupted, else 'F'.<br></td>
         </tr>
         <tr>
            <td><a href="DataSet.html">DataSetWID</a></td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>No</td>
            <td>Reference to the data set from which the entity came from<br></td>
         </tr>
      </table>
      <h2>Referenced By</h2>
      <table border="1">
         <tr>
            <th>Table</th>
            <th>Column</th>
         </tr>
         <tr>
            <td><a href="BioSourceWIDGeneWID.html">BioSourceWIDGeneWID</a></td>
            <td>GeneWID</td>
         </tr>
         <tr>
            <td><a href="GeneWIDNucleicAcidWID.html">GeneWIDNucleicAcidWID</a></td>
            <td>GeneWID</td>
         </tr>
         <tr>
            <td><a href="GeneWIDProteinWID.html">GeneWIDProteinWID</a></td>
            <td>GeneWID</td>
         </tr>
         <tr>
            <td><a href="Entry.html">Entry</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="TranscriptionUnitComponent.html">TranscriptionUnitComponent</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="Support.html">Support</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="RelatedTerm.html">RelatedTerm</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CitationWIDOtherWID.html">CitationWIDOtherWID</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CommentTable.html">CommentTable</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CrossReference.html">CrossReference</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CrossReference.html">CrossReference</a></td>
            <td>CrossWID</td>
         </tr>
         <tr>
            <td><a href="Description.html">Description</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="DBID.html">DBID</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="SynonymTable.html">SynonymTable</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="ToolAdvice.html">ToolAdvice</a></td>
            <td>OtherWID</td>
         </tr>
      </table>
      <h2>Other Constraints</h2>None.
      <h2>Indexes</h2>
      <table border="1">
         <tr>
            <th>Name</th>
            <th>Columns</th>
         </tr>
         <tr>
            <td>GENE_DATASETWID</td>
            <td>DATASETWID</td>
         </tr>
         <tr>
            <td>GENE_Name</td>
            <td>Name</td>
         </tr>
         <tr>
            <td>GENE_Type</td>
            <td>Type</td>
         </tr>
         <tr>
            <td>GENE_GenomeID</td>
            <td>GenomeID</td>
         </tr>
         <tr>
            <td>GENE_START_END_POSITION</td>
            <td>CODINGREGIONSTART, CODINGREGIONEND</td>
         </tr>
         <tr>
            <td>GENE_ENDPOSITION</td>
            <td>CODINGREGIONEND</td>
         </tr>
         <tr>
            <td>GENE_NAWID</td>
            <td>NucleicAcidWID</td>
         </tr>
         <tr>
            <td>GENE_SubsequenceWID</td>
            <td>SubsequenceWID</td>
         </tr>
      </table><br><br></body>
</html>