<html>
   <head>
      <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
      <link href="style.css" rel="stylesheet" type="text/css">
      <title>SchemaDoc:
         NucleicAcid
      </title>
   </head>
   <body><a href="index.html">Table Index</a><h1><a name="NucleicAcid"></a>NucleicAcid
      </h1>
      <hr>Defines a specific nucleic acid molecule, such as DNA or of RNA.<br>Entries in this table correspond to a real-world DNA or RNA molecules<br>which at one point were purified and isolated ("something you can point at"). A row in this
      table
      <br>can define a complete molecule, a fragment of a molecule, or a molecule that has been
      partially
      <br>sequenced in different regions.<br>This table will be used in several ways: (1) to associate a sequence with an entire<br>replicon, or a region of a replicon, when the sequence of that replicon is known; (2) to
      associate
      <br>a DNA sequence with a single gene; (3) to define a DNA or RNA molecule that has not been<br>sequenced; (4) to define RNA molecules such as tRNAs.<br>Features on a NucleicAcid molecule (such as promoters or binding sites) can be defined using<br>the Feature table.<br>The Subsequence table contains zero or more full or partial sequences contents of this
      molecule.
      <br>If one represents the full sequence of this molecule, Subsequence.FullSequence = 'T' where<br>Subsequence.NucleicAcidWID references this molecule. In this case, NucleicAcid.FullySequenced = 'T' as
      well.
      <br><h2>Columns</h2>
      <table border="1">
         <tr>
            <th>Column</th>
            <th>MySQL Type</th>
            <th>Oracle Type</th>
            <th>Nullable</th>
            <th>Description</th>
         </tr>
         <tr>
            <td>WID</td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>No</td>
            <td>Warehouse identifier for this NucleicAcid.<br></td>
         </tr>
         <tr>
            <td>Name</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(200)</td>
            <td>Yes</td>
            <td>Name or description of this molecule.<br>Ex: (CMR) 'Chromosome II Brucella melitensis 16M'.<br></td>
         </tr>
         <tr>
            <td>Type</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(30)</td>
            <td>No</td>
            <td>Enumeration: 'DNA' 'RNA' or 'na'<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>DNA</b> -
                  The molecule is composed of DNA<br><b>RNA</b> -
                  The molecule is composed of RNA<br><b>NA</b> -
                  The molecule is specified as a nucleic acid but whether of type DNA or RNA is not known<br></div>
            </td>
         </tr>
         <tr>
            <td>Class</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(30)</td>
            <td>Yes</td>
            <td>Enumeration: describes the molecule as it exists in the organism; stores values such as RNA subtype
               ("pre-RNA", "mRNA"), replicon type (e.g., "chromosome", "plasmid")
               <br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>pre-RNA</b> -
                  As it exists within the organism, the molecule is a pre-RNA molecule. NCBI: used when there is no evidence that mature RNA
                  is produced<br><b>mRNA</b> -
                  As it exists within the organism, the molecule is a mature mRNA. NCBI: used when there is evidence that mature mRNA is produced<br><b>rRNA</b> -
                  As it exists within the organism, the molecule is an rRNA; NCBI used when there is evidence that mature rRNA is produced<br><b>tRNA</b> -
                  As it exists within the organism, the molecule is a tRNA; NCBI: used when there is evidence that mature tRNA is produced<br><b>snRNA</b> -
                  As it exists within the organism, the molecule is a small nuclear RNA. NCBI: used when there is evidence that snRNA is produced<br><b>scRNA</b> -
                  As it exists within the organism, the molecule is an RNA which encodes small cytoplasmic ribonucleic proteins. NCBI: used
                  when there is evidence that gene codes of small cytoplasmic (sc) ribonucleoproteins (RNP)s<br><b>snoRNA</b> -
                  As it exists within the organism, the molecule is a small nucleolar RNA. NCBI: used when there is evidence that transcript
                  is a small nucleolar RNA<br><b>other</b> -
                  Check notes as to how we map this...<br><b>chromosome</b> -
                  As it exists within the organism, the molecule is a chromosome<br><b>plasmid</b> -
                  As it exists within the organism, the molecule is a plasmid<br><b>organelle-chromosome</b> -
                  As it exists within the organism, the molecule is the chromosome of an organelle<br><b>transposon</b> -
                  As it exists within the organism, the molecule is a transposon<br><b>virus</b> -
                  As it exists within the organism, the molecule is a virus<br><b>unknown</b> -
                  Used when the class is unknown<br></div>
            </td>
         </tr>
         <tr>
            <td>Topology</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(30)</td>
            <td>Yes</td>
            <td>Enumeration: 'circular', 'linear' or 'other'.<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>linear</b> -
                  The topology of the molecule is Linear<br><b>circular</b> -
                  The topology of the molecule is Circular<br><b>other</b> -
                  The topology of the molecule is neither circular nor linear but is known<br></div>
            </td>
         </tr>
         <tr>
            <td>Strandedness</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(30)</td>
            <td>Yes</td>
            <td>Enumeration indicating whether Nucleic Acid is single stranded, double stranded or mixed
               stranded.
               <br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>ss</b> -
                  The molecule is single stranded<br><b>ds</b> -
                  The molecule is double stranded<br><b>mixed</b> -
                  The molecule is composed of single and double stranded regions<br></div>
            </td>
         </tr>
         <tr>
            <td>SequenceDerivation</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(30)</td>
            <td>Yes</td>
            <td>Enumeration describing how the sequence was generated<br>(e.g., from single clone, from assembly of sequences<br>from other sequence entries, collection of different clones, etc).<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>virtual</b> -
                  NCBI: The sequence of the molecule is NOT known. NOTE: this should map to null.<br><b>raw</b> -
                  POORLY DEFINED; seems to be used to indicate that the sequence of the molecule was actually generated from one single continuous
                  molecule, as opposed to assembled together from sequences from different molecules (e.g., different clones)<br><b>seg</b> -
                  NCBI: The sequence of the molecule is made up of collection of segments arranged according to specified coordinates, e.g.,
                  sequence was derived from assembling sequences from different clones<br><b>reference</b> -
                  NCBI: The sequence of the molecule is constructed from existing Bioseqs;It behaves exactly like a segmented Bioseq in taking
                  it's data and character from the Bioseq to which it points.<br><b>constructed</b> -
                  NCBI: The sequence of the molecule is constructed by assembling other Bioseqs<br><b>consensus</b> -
                  NCBI: The sequence of the molecule represents a pattern typical of a sequence region or family of sequences;' it summarizes
                  attributes of an aligned collection of real sequences. Note that this is NOT A REAL OBJECT<br><b>map</b> -
                  NCBI: The molecule does not have a sequence describing it, but rather a set of coordinates (restriction fragment order, genetic
                  markers, physical map, etc)<br></div>
            </td>
         </tr>
         <tr>
            <td>Fragment</td>
            <td>CHAR(1)</td>
            <td>CHAR(1)</td>
            <td>Yes</td>
            <td>'T' if this is a fragment of a molecule,<br>'F' if this NucleicAcid describes an entire molecule.<br></td>
         </tr>
         <tr>
            <td>FullySequenced</td>
            <td>CHAR(1)</td>
            <td>CHAR(1)</td>
            <td>Yes</td>
            <td>'T' if the molecule is completely<br>sequenced within this dataset, else 'F'.<br></td>
         </tr>
         <tr>
            <td>MoleculeLength</td>
            <td>INT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>The length of the molecule, in nucleotides. This value is firm if FullySequenced is 'T';<br>otherwise,MoleculeLength is an approximation and is not necessarily equal to
               CumulativeLength,
               <br>as the the molecule may not have been sequenced to completion.<br>This value is calculated if it is not explicitely stated in the source dataset<br></td>
         </tr>
         <tr>
            <td>MoleculeLengthApproximate</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(10)</td>
            <td>Yes</td>
            <td>Enumeration; specifies whether MoleculeLength stores an approximate value:<br>'gt' for greater than,<br>'lt' for less than, or<br>'ne' for not equal.<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>gt</b> -
                  The length of the molecule's sequence is greater than the actual length specified<br><b>lt</b> -
                  The length of the molecule's sequence is less than the actual length specified<br><b>ne</b> -
                  The length of the molecule's sequence is less than or greater than the actual length. All we know is that its not the exact
                  length<br></div>
            </td>
         </tr>
         <tr>
            <td>CumulativeLength</td>
            <td>INT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>The cumulative number of nucleotides for all Subsequences referenced by this NucleicAcid
               entry,
               <br>whether contiguous or not. This value is a summation of the number of nucleotides for these
               Subsequences.
               <br>If the molecule is completely sequenced, this value should be identical<br>to that of MoleculeLength, and both fields are populated.<br></td>
         </tr>
         <tr>
            <td>CumulativeLengthApproximate</td>
            <td>VARCHAR</td>
            <td>VARCHAR2(10)</td>
            <td>Yes</td>
            <td>Enumeration; specifies whether CumulativeLength how approximates actual total length:<br>'gt' for greater than,<br>'lt' for less than, or<br>'ne' for not equal.<br><br><div style="margin-left: 40px;"><b><i>Enumerated Values:</i></b><br><b>gt</b> -
                  The total length of the molecule's sequence is greater than the actual length<br><b>lt</b> -
                  The total length of the molecule's sequence is less than the actual length<br><b>ne</b> -
                  The total length of the molecule's sequence is less than or greater than the actual length. All we know is that its not the
                  exact length<br></div>
            </td>
         </tr>
         <tr>
            <td><a href="GeneticCode.html">GeneticCodeWID</a></td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>References the genetic code of this molecule.<br></td>
         </tr>
         <tr>
            <td><a href="BioSource.html">BioSourceWID</a></td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>Yes</td>
            <td>References the biological source of this molecule.<br></td>
         </tr>
         <tr>
            <td><a href="DataSet.html">DataSetWID</a></td>
            <td>BIGINT</td>
            <td>NUMBER</td>
            <td>No</td>
            <td>References the data set from which the entity came from<br></td>
         </tr>
      </table>
      <h2>Referenced By</h2>
      <table border="1">
         <tr>
            <th>Table</th>
            <th>Column</th>
         </tr>
         <tr>
            <td><a href="Subsequence.html">Subsequence</a></td>
            <td>NucleicAcidWID</td>
         </tr>
         <tr>
            <td><a href="Gene.html">Gene</a></td>
            <td>NucleicAcidWID</td>
         </tr>
         <tr>
            <td><a href="GeneWIDNucleicAcidWID.html">GeneWIDNucleicAcidWID</a></td>
            <td>NucleicAcidWID</td>
         </tr>
         <tr>
            <td><a href="Entry.html">Entry</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="Support.html">Support</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="InteractionParticipant.html">InteractionParticipant</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="RelatedTerm.html">RelatedTerm</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CitationWIDOtherWID.html">CitationWIDOtherWID</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CommentTable.html">CommentTable</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CrossReference.html">CrossReference</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="CrossReference.html">CrossReference</a></td>
            <td>CrossWID</td>
         </tr>
         <tr>
            <td><a href="Description.html">Description</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="DBID.html">DBID</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="SynonymTable.html">SynonymTable</a></td>
            <td>OtherWID</td>
         </tr>
         <tr>
            <td><a href="ToolAdvice.html">ToolAdvice</a></td>
            <td>OtherWID</td>
         </tr>
      </table>
      <h2>Other Constraints</h2>None.
      <h2>Indexes</h2>
      <table border="1">
         <tr>
            <th>Name</th>
            <th>Columns</th>
         </tr>
         <tr>
            <td>NUCLEICACID_Name</td>
            <td>Name</td>
         </tr>
         <tr>
            <td>NUCLEICACID_Type</td>
            <td>Type</td>
         </tr>
         <tr>
            <td>NUCLEICACID_Class</td>
            <td>Class</td>
         </tr>
         <tr>
            <td>NUCLEICACID_MoleculeLength</td>
            <td>MoleculeLength</td>
         </tr>
         <tr>
            <td>NUCLEICACID_CumulativeLength</td>
            <td>CumulativeLength</td>
         </tr>
         <tr>
            <td>NUCLEICACID_GCWID</td>
            <td>GeneticCodeWID</td>
         </tr>
         <tr>
            <td>NUCLEICACID_BSWID</td>
            <td>BioSourceWID</td>
         </tr>
         <tr>
            <td>NUCLEICACID_DATASETWID</td>
            <td>DATASETWID</td>
         </tr>
      </table><br><br></body>
</html>