<!DOCTYPE html PUBLIC "-//w3c//dtd html 4.0 transitional//en">
<html>
<head>
  <meta http-equiv="Content-Type"
 content="text/html; charset=iso-8859-1">
<!-- <link rel="stylesheet" href="themes/SeaBreeze/style/style.css" type="text/css"> -->
  <title>BioCyc Loader for BioWarehouse</title>
</head>
<body style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);">
<!-- <img src="images/RHLogo.jpg" width="400" height="90"> -->
<p>&nbsp;</p>
<center>
<h1>BioCyc Loader for BioWarehouse</h1>
<h2>Version 4.6</h2>
<br>
</center>
<!-- typ- and tag-ography for

  files: natural case, <KBD>

  warehouse tables and columns: capitalized, <code>

  input file attributes: uppercase, bold font

-->
<div style="text-align: center;"><code></code><code>(C) 2006 SRI
International. All Rights Reserved.&nbsp; See <a
 href="../../../doc/index.html">BioWarehouse
Overview</a> for license details.</code></div>

<pre><hr><br><br></pre>
<a href="#Intro">Introduction</a><br>
<a href="#Limitations">Limitations</a><br>
<a href="#InstallBuild">Installation and Building</a><br>
<a href="#Input">Obtaining input data</a><br>
<a href="#Dependencies">Loader Dependencies and Prerequisites</a><br>
<a href="#Run">Running the Loader</a><br>
<a href="#Dataset">Dataset and BioSource Specification</a><br>
<a href="#ObjectTranslation">Translation Semantics for BioCyc Objects</a><br>
<LI><a href="#Publications">Publications</a><br>
<li><a href="#Compounds">Compounds</a><br>
</li>
<li><a href="#Proteins">Proteins</a><br>
</li>
<li><a href="#ProtSeq">Protein Sequences</a><br>
</li>
<li><a href="#TransUnits">Transcription Units</a><br>
</li>
<li><a href="#Genes">Genes</a><br>
</li>
<li><a href="#Promoters">Promoters</a><br>
</li>
<li><a href="#Terminators">Terminators</a><br>
</li>
<li><a href="#DNABindsites">DNA Binding Sites</a><br>
</li>
<li><a href="#Reactions">Reactions</a><br>
</li>
<li><a href="#EnzRxns">Enzymatic Reactions</a><br>
</li>
<li><a href="#Regulation">Regulation</a><br>
</li>
<li><a href="#Pathways">Pathways</a><br>
</li>
<li><a href="#SupportTable">Support Table</a><br>
</li>
<li><a href="#EntryTable">Entry Table</a><br>

<a href="#Reference">References</a><br>

  <hr> <a name="Intro"></a>
  <h1><a name="Intro">Introduction</a></h1>
  <a name="Intro"> </a>
  <p><a name="Intro">This document describes version 4.6 of
the BioCyc
Loader. It is one of several database loaders comprising the BioWarehouse.
The BioCyc Loader (referred to simply as <em>the loader</em>), loads a
  </a><a href="http://bioinformatics.ai.sri.com/ptools/">Pathway/Genome
Database (PGDB)</a> into the BioWarehouse - a relational
database that provides a common representation for diverse
bioinformatics databases. </p>
  <p> PGDBs are implemented in a frame-based representation system
which is implemented in Common Lisp. The loader inputs a textual flat
file representation of a PGDB, converts it to the representation
expressed in the <a href="../../../doc/Schema.html"> Bio-SPICE
Warehouse Schema</a>,
and loads this directly into an instance of the warehouse. </p>
  <h2>Overview of BioWarehouse Schema</h2>
The Bio-SPICE warehouse schema
contains the data definition statements for the BioWarehouse.
These include three different types of tables - <em> constant tables</em>,
  <em> object tables</em>, <em> linking tables</em>, and <em> special
tables</em>.
  <p>Constant tables specify scientific data such as information from
the Periodic Table of Elements, as well as constants used as column
values in various warehouse tables. </p>
  <p>Object tables describe a type of entity in a source database, such
as compounds and proteins.
Each column of an object table specifies a parameter that characterizes
the object.
In addition to the parameters defined by the source database, the
loader assigns a unique
em&gt; warehouse ID (WID)
to each object, which is used by other tables to reference the object. </p>
  <p>A special type of warehouse object is the <em> dataset</em>. A
dataset object is created for each dataset
loaded into the warehouse, i.e., the SWISS-PROT loader adds one row to
this table when it is run. Its WID is referred to as the <em> dataset
WID</em>
and is a column in each object table, specifying the source database of
the object. </p>
  <p>A linking table describes relationships among objects. They
contain WIDs of the associated objects, and any additional columns
needed to characterize the relationship. In general, many-to-many
relationships are supported.
Special tables exist to capture reference and crossreference
information and to facilitate lookup of objects. </p>
  <p><a href="../../../doc/Schema.html">Schema
documentation</a> is available.</p>

  <hr>
  <h1><a name="Limitations"></a>Limitations</h1>
  <p>The latest supported data version for the BioCyc loader is listed
in the <a href="../../../doc/index.html#loaders">loader summary table</a>.&nbsp;
Attributes added to the BioCyc schema after this version are not
supported. </p>
 <P>
The loader silently ignores the numerous source attributes that have no analogue in the BioWarehouse.
 <P>
In the reaction graph specification in the <B>PREDECESSORS</B> attribute of pathways,
due to parsing limitations a reaction may have at most two predecessor reactions.
Additional reactions are flagged as generic syntax errors.
  <p>The loader treats the <kbd>MetaCyc</kbd> database as if contained data for a single organism
  named "MetaCyc", rather than as containing experimentally elucidated data from many organisms. </p>
 <P>
Many proteins do not have the  <B>COMMON-NAME</B> attribute.  This is because
multifunctional enzymes often have different names depending on which
enzymatic function is being referred to.   If no common name is
specified (and we prefer that no common name be specified if an
activity name is appropriate), we use the common name of the
corresponding enzymatic-reaction frame (or a concatenation of them
separated by / if there there are multiple enzymatic-reactions).
 <P>
The individual components of a publication -- title, year, etc. -- are not
broken out by the loader into the associated columns of the  <code>Citation</code>
table; only the concatenation of all components is loaded into  <code>Citation.Citation</code>.
 <P>
The TRANSCRIPTION-DIRECTION attribute is ignored for transcription units.
In particular, neither it nor the TRANSCRIPTION-DIRECTION of the associated gene is
used in computing Feature.StartPosition of Feature.EndPosition for DNA binding sites.
This could lead to incorrect values for these columns.

  <hr><a name="InstallBuild"> </a>
  <h1><a name="InstallBuild">Installation and Building</a></h1>
  <a name="InstallBuild">See </a><a
 href="../../../biocyc-loader/flatfile/doc/index.html">BioCyc
installation instructions</a>
for details on installing and building the loader.

  <hr><a name="Dependencies"> </a>
  <h1><a name="Dependencies">Loader Dependencies and Prerequisites</a></h1>
  <a name="Dependencies">Other that the standard warehouse creation
procedure,
the loader does not require that any other Warehouse tools be run prior
to its execution. However,
if the </a><a
 href="../../../metacyc-ontology-loader/doc/metacyc-ontology-manual.html">
MetaCyc Ontology</a>
or <a href="../../../go-loader/docs/go-manual.html">Gene Ontology</a>
datasets have been loaded, the loader builds useful links to objects in these
datasets.  A warning is issued if either of these datasets has not been
loaded. If multiple versions of these datasets are present in the Warehouse,
the dataset with the maximum WID (typically the most recently loaded) is
used.
<p>The Multifun Ontology contained in the MetaCyc Ontology dataset is used to populate rows of the
  <code>RelatedTerm</code> table for genes that specify a Multifun type.
The Gene Ontology dataset is used to populate rows of the
  <code>RelatedTerm</code> table for proteins that specify a GO term.

See <a href="#Genes">genes translation</a>
and  <a href="#Proteins">proteins translation</a>
for details. </p>

<p>The loader is invoked as part of the <A HREF="../../../chip-chip-loader/docs">ChIP-chip loader</A>. When invoked in this manner, a restricted set of data files, with a minimal amount of data, are loaded and merged into a dataset containing other BioWarehouse data as well. Also, the command-line arguments provided to the loader are specified in a properties file. See the ChIP-chip documentation for full details.</p>


  <hr><a name="Input"> </a>
  <h1><a name="Input">Input data</a></h1>
  <a name="Input">BioCyc PGDB databases are available for a number of
species.
However, a license may be required to obtain them. Visit </a><a
 href="http://biocyc.org/download.shtml">BioCyc downloads</a>
or send a request to <a href="mailto:biocyc-info@ai.sri.com">biocyc-info@ai.sri.com</a>
for details.
  <p>The textual representation of a PGDB consists of several ASCII
files. A subset of these are used by the BioCyc loader. Furthermore,
not all files are present in all PGDBs; in particular, several files
are not present for the MetaCyc PGDB.
Input files are loaded in the following order: </p>
  <ol>
    <code><li> <kbd>pubs.dat</kbd>  [not present for all BioCyc PDDBs] </li>
    <li> <kbd>compounds.dat</kbd> </li>
    <li> <kbd>proteins.dat</kbd> </li>
    <li> <kbd>protseq.fasta</kbd> [not present for all BioCyc PGDBs] </li>
    <li> <kbd>transunits.dat</kbd>  [not present for all BioCyc PGDBs] </li>
    <li> <kbd>genes.dat</kbd> </li>
    <li> <kbd>promoters.dat</kbd> [not present for the MetaCyc PGDB] </li>
    <li> <kbd>terminators.dat</kbd> [not present for the MetaCyc PGDB] </li>
    <li> <kbd>dnabindsites.dat</kbd> [not present for the MetaCyc PGDB] </li>
    <li> <kbd>reactions.dat</kbd> </li>
    <li> <kbd>enzrxns.dat</kbd> </li>
    <li> <kbd>regulation.dat</kbd> </li>
    <li> <kbd>pathways.dat</kbd> </li>
    </code>
  </ol>
  <p> See the <a
 href="http://bioinformatics.ai.sri.com/ptools/flatfile-format.html">
PGDB flat file format specification</a>
for detailed specification on the contents of the input files used by
the loader. Most files are in <a
 href="http://bioinformatics.ai.sri.com/ptools/flatfile-format.html#Attribute-Value">attribute-value</a>
format. </p>

  <hr><a name="Run"> </a>
  <h1><a name="Run">Running the Loader</a></h1>
  <a name="Run">The </a><a
 href="../../../biocyc-loader/flatfile/doc/index.html">BioCyc
installation instructions</a>
contain details for running the loader, including options and a
description of its output.

  <hr><a name="Dataset"> </a>
  <h1><a name="Dataset">Dataset and BioSource Specification</a></h1>
  <a name="Dataset">If the -m (merge) command line option is used, the
loader loads data into the dataset named "BioCyc",
using the WID of this entry as the DataSetWID for all objects it adds
to the Warehouse.
If multiple datasets of this name exist, the one with the
maximal <code>DataSet.WID</code> is used; typically this is the
dataset that was most recently loaded.
If no dataset of this name exists, a warning is issued and one is
created. </a>
  <p><a name="Dataset">If the -m command line option is not used,
the loader adds one row to the <code>Dataset</code> table as follows,
using the WID of this entry
as the DataSetWID for all objects it adds to the Warehouse: </a></p>
  <p>
  <table border="5">
    <caption align="top"><b>Column values for <code>Dataset</code> row</b></caption>
    <tbody>
      <tr>
        <th align="left">Column</th>
        <th align="left">Value assigned by BioCyc loader</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <td align="left"><code>WID</code> </td>
        <td align="left"> A small integer that uniquely identifies this
dataset in the warehouse. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Name</code> </td>
        <td align="left"> '<i>species</i>Cyc', where <i>species</i> is
an abbreviation for the organism represented in the PGDB. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Version</code> </td>
        <td align="left"> Major version of PGDB that is loaded. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>ReleaseDate</code> </td>
        <td align="left"> The date that this version of the PGDB was released. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>LoadDate</code> </td>
        <td align="left"> The date and time the loader was started. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>ChangeDate</code> </td>
        <td align="left"> The date and time the loader completed, NULL if the loader did not complete successfully. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>LoadedBy</code> </td>
        <td align="left"> The value of the system environment variable <b>USER</b>
for the account running the loader. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Application</code> </td>
        <td align="left"> 'BioCyc Loader' </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>ApplicationVersion</code> </td>
        <td align="left"> 4.6 </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>HomeURL</code> </td>
        <td align="left"> <a href="http://www.biocyc.org">http://www.biocyc.org</a>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>QueryURL</code> </td>
        <td align="left"> <a href="http://www.biocyc.org:1555">http://www.biocyc.org:1555</a>
        </td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </p>
  <p> The loader adds one row to the <code>BioSource</code> table as
follows, defining the organism the BioCyc KB describes: </p>
  <p>
  <table border="5">
    <caption align="top"><b>Column values for <code>BioSource</code>
row</b></caption> <tbody>
      <tr>
        <th align="left">Column</th>
        <th align="left">Value assigned by BioCyc loader</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <td align="left"><code>WID</code> </td>
        <td align="left"> The next available WID in the warehouse.
Uniquely specifies this BioSource in the warehouse. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Name</code> </td>
        <td align="left"> The value specified by the <kbd>-o</kbd>
command line option, e.g., 'Bacillus subtilis' </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Strain</code> </td>
        <td align="left"> NULL </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>DatasetWID</code> </td>
        <td align="left"> The value <code>Dataset.WID</code> assigned
to the dataset being loaded. </td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </p>

  <hr><a name="ObjectTranslation"> </a>
  <h1><a name="ObjectTranslation">Translation Semantics for BioCyc
Objects</a></h1>
  <a name="ObjectTranslation"> </a>
  <p><a name="ObjectTranslation">This section describes the semantic
mapping between the objects comprising a BioCyc knowledge base
and its associated flat file representation to its representation in
the Bio-SPICE
data warehouse. Semantics are expressed in tabular form, showing the
mapping of each source attribute to the warehouse <code>Table.Column</code>
values computed from it. The most typical case is that the attribute is
simply copied into a warehouse column; if translation is more complex,
an explanation is given. Any attributes not listed are ignored. </a></p>
  <p><a name="ObjectTranslation">Some attributes can occur multiple
times for a source object. The notation <b>ATTRIBUTE[*]</b> is used to
indicate that the semantics apply to all occurrences; typically a row
is added to a warehouse table for each. The notation <b>ATTRIBUTE[1]</b>,
  <b>ATTRIBUTE[2]</b>, etc., is used where the attribute order is
significant.
If an attribute is missing from a source file but required by the
warehouse schema (i.e., its column is qualified with NOT NULL), a
warning is issued. If the missing attribute is not required, NULL is
stored. </a></p>
  
  <p><a name="Publications"> </a></p>
  <h2><a name="Publications">Publications</a></h2>
  <a name="Publications">Publications are input from the file <kbd>pubs.dat</kbd>.
Any BioCyc object can contain a reference to a publication. This is
done with the <b>CITATIONS</b> attribute. This attribute contains
the <b>UNIQUE-ID</b> of a publication, possibly enclosed in square
brackets, and possibly without the leading "PUB-" prefix.
Certain citations do not refer to publications; they refer to evidence.
See the  <a href="#SupportTable">Support Table</a> for translation details. </a>

  <p><a name="Publications">A row is added to the <code>Citation</code>
table for each entry in <kbd>pubs.dat</kbd>, except for those with a <b>REFERENT-FRAME</b>
attribute. The latter provide an alternative name for a publication -
either the
publication's <b>UNIQUE-ID </b> or its <b>REFERENT-FRAME</b>
attribute may be used to refer to the publication in <b>CITATIONS</b>
attributes in other files. </a></p>
  <p><a name="Publications"><br>
  </a><a name="pubs.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>pubs.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">AUTHOR[*]</th>
        <td align="left"> Concatenated together to form the list of
authors of the publication. The order of the authors is preserved.<br>
A comma and a space are inserted between each author. <br>
Along with other attributes, this is included in the full text of the
citation, stored at <code>Citation.Citation</code>.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Citation</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">MEDLINE-UID</th>
        <td align="left"> <code>Crossreference.XID</code>; <br>
        <code>Crossreference.DatabaseName</code> is 'Medline'; <br>
        <code>Crossreference.OtherWID</code> is the WID of this <code>Citation</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PUBMED-ID</th>
        <td align="left"> <code>Citation.PMID</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REFERENT-FRAME</th>
        <td align="left">Attribute provides an alternate name for <b>UNIQUE-ID</b>;
either may be used in other files to refer to the publication.<br>
Both refer to the same <code>Citation</code> object. <br>
Used internally to associate a publication name with a <code>Citation.WID</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">TITLE, <br>
SOURCE, <br>
YEAR, <br>
URL</th>
        <td align="left"> Concatenated together in the given order and
appended to the full list of authors derived from the <b>AUTHOR[*]</b>
attributes <br>
to form <code>Citation.Citation</code>. Tabs are inserted between each
attribute and the between the author list and <b>TITLE</b>. <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Citation</code>
object. <br>
Used internally to associate a publication name with a <code>Citation.WID</code>.</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>
No additional table rows are added for publications.

<a name="Compounds"></a>
  <h2><a name="Compounds">Compounds</a></h2>
  <a name="Compounds">Compounds are input from the file <kbd>compounds.dat</kbd>.
A row is added to the <code>Chemical</code> table for each entry in <kbd>compounds.dat</kbd>.
A <code>Chemical</code> may be either a single compound or a class of compounds; all
entries from this file are single compounds; hence <code>Chemical.Class</code> is always 'F'.
 <P>
Note that rows are also added to <code>Chemical</code> when translating <KBD>reactions.dat</KBD>;
these compounds are classes of compounds, and their <code>Chemical.Class</code> is 'T'.
  </a>
  <p><a name="Compounds"><br>
  </a><a name="compounds.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>compounds.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CAS-REGISTRY-NUMBERS</th>
        <td align="left"> <code>Chemical.CAS</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">CHARGE</th>
        <td align="left"> <code>Chemical.Charge</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">CHEMICAL-FORMULA[*]</th>
        <td align="left"> All (Element Number) pairs are concatenated
to form <code>Chemical.EmpiricalFormula</code> <br>
Ex: (H 2) and (O 1) form 'H2O1'</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Chemical</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Chemical</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Chemical.Name</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Chemical</code>
object; <br>
        <code>CrossReference.DatabaseName</code> is the first element
of the list; <br>
        <code>CrossReference.XID</code> is the second element of the
list; <br>
the rest of the list is ignored. </td>
      </tr>
      <tr>
        <th align="left">MOLECULAR-WEIGHT</th>
        <td align="left"> <code>Chemical.MolecularWeightCalc</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PKA1</th>
        <td align="left"> <code>Chemical.PKA1</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PKA2</th>
        <td align="left"> <code>Chemical.PKA2</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PKA3</th>
        <td align="left"> <code>Chemical.PKA3</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REGULATES</th>
        <td align="left"> Ignored. This is the converse of the <B>REGULATOR</B> attribute for an enzymatic <A HREF="#Regulation">regulator</A>. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SMILES</th>
        <td align="left"> <code>Chemical.Smiles</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Chemical</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYSTEMATIC-NAME</th>
        <td align="left"> <code>Chemical.SystematicName</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Chemical</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">ACTIVATORS-UNMECH-OF, <BR>
                         INHIBITORS-ALLOSTERIC-OF, <BR>
                         INHIBITORS-IRRREVERSIBLE-OF, <BR>
                         INHIBITORS-OTHER-OF
	</th>	
        <td align="left"> Ignored; these are symmetric analogues to the corresponding attributes <B>INHIBITORS-ALLOSTERIC</B>, etc. of <KBD>enzrxns.dat</KBD>.
	</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>
No additional table rows are added for compounds.

<a name="Proteins"></a>
  <h2><a name="Proteins">Proteins</a></h2>
  <a name="Proteins">Proteins are input from the file <kbd>proteins.dat</kbd>.
A row is added to the <code>Protein</code> table for each entry in <kbd>proteins.dat</kbd>.
  </a>
  <p><a name="Proteins"><br>
  </a><a name="proteins.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>proteins.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Protein</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">^COEFFICIENT[*]</th>
        <td align="left"> <code>Subunit.Coefficient</code> for the
immediately preceding <b>COMPONENTS</b> attribute. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Protein.Name</code>. <BR>
Many proteins do not have the <B>COMMON-NAME</B> attribute.  This is because
multifunctional enzymes often have different names depending on which
enzymatic function is being referred to.   If no common name is
specified (and we prefer that no common name be specified if an
activity name is appropriate), we use the common name of the
corresponding enzymatic-reaction frame (or a concatenation of them
separated by / if there there are multiple enzymatic-reactions).
	</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Protein</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENTS[*]</th>
        <td align="left"> Attribute is a <b>UNIQUE-ID</b> for a
protein. A row is added to <code>Subunit</code>; <br>
        <code>Subunit.ComplexWID</code> is the WID of this <code>Protein</code>
object; <br>
        <code>Subunit.SubunitWID</code> is the WID of the protein
associated with the attribute; <br>
        <code>Subunit.Coefficient</code> is the value of the
immediately following <b>^COEFFICIENT</b> attribute, <br>
defaulting to 1 if not explicit <br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Protein</code>
object; <br>
        <code>CrossReference.DatabaseName</code> is the first element
of the list; <br>
        <code>CrossReference.XID</code> is the second element of the
list; <br>
the rest of the list is ignored. </td>
      </tr>
      <tr>
        <th align="left">FEATURES</th>
        <td align="left"> Ignored</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">GENE</th>
        <td align="left"> Ignored</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">GO-TERMS[*]</th>
        <td align="left"> Ignored unless the  <A HREF="../../../go-loader/docs/go-manual.html">Gene Ontology</A> dataset has been previously loaded; each term is a <CODE>DBID.XID</CODE> from that dataset. <BR>
	A row is added to <code>RelatedTerm</code>: <br>
        <code>RelatedTerm.TermWID</code> references this term; <br>
        <code>RelatedTerm.OtherWID</code> is the WID of this <code>Protein</code> object. <BR>
        <code>RelatedTerm.Relationship</code> is 'keyword'</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">LOCATION[*]</th>
        <td align="left"> <code>Location.Location</code>; <br>
        <code>Location.ProteinWID</code> is the WID of this <code>Protein</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
      <tr>
        <th align="left">MOLECULAR-WEIGHT-SEQ</th>
        <td align="left"> <code>Protein.MolecularWeightCalc</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">MOLECULAR-WEIGHT-EXP</th>
        <td align="left"> <code>Protein.MolecularWeightExp</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PI</th>
        <td align="left"> <code>Protein.PICalc</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Protein</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Protein</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">ACTIVATORS-UNMECH-OF, <BR>
                         COFACTORS-UNMECH-OF, <BR>
                         INHIBITORS-UNMECH-OF, <BR>
                         INHIBITORS-COMPETITIVE-OF, <BR>
                         INHIBITORS-OTHER-OF, <BR>
                         PROSTHETIC-GROUPS-OF
	</th>	
        <td align="left"> Ignored; these are symmetric analogues to the corresponding attributes <B>INHIBITORS-COMPETITIVE</B>, etc. of <KBD>enzrxns.dat</KBD>.
	</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>
A row is added to <code>BioSourceWIDProteinWID</code> for each protein
entry. <code>BioSourceWIDProteinWID.BioSourceWID</code> is the WID
from the one row of <code>BioSource</code> created by the loader for
the species being loaded.
Numerous other table rows are added for proteins, but they are added
when the linked object is parsed. In particular, the <b>GENE</b>
attribute is ignored, and the protein - gene link is created when genes
are parsed.
 <P>
 <P>
If the MetaCyc Ontology dataset has been loaded,
a row is added to <CODE>RelatedTerm</CODE> for each <B>GO-TERMS</B> attribute that is a term
in <A HREF="../../../go-loader/docs/go-manual.html">Gene Ontology</A>.

<a name="ProtSeq"></a>
  <h2><a name="ProtSeq">Protein Sequences</a></h2>
Protein sequences are input from the file
protseq.fasta. This file is not an attribute-value format file; each
entry contains a protein name, some other information which is ignored,
and an amino acid sequence. The row for that protein is updated to
store the amino acid sequence as <code>Protein.AASequence</code>.

  <h3><a name="ProtSeq">Additional Tables</a></h3>
  <a name="ProtSeq">No additional table rows are added based on entries
in protseq.fasta. </a>


<h2><a name="TransUnits">Transcription Units</a></h2>
  <a name="TransUnits">Transcription Units are input from the file <kbd>transunits.dat</kbd>.
A row is added to the <code>TranscriptionUnit</code> table for each entry in this file.
  </a>
  <p><a name=">Transcription Units"><br>
  </a><a name="transunits.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>transunits.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code> associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>TranscriptionUnit</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>TranscriptionUnit</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>TranscriptionUnit.Name</code><br>
	If missing, <B>UNIQUE-ID</B> is used in its place.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENTS[*]</th>
        <td align="left"> Ignored; a link to this entry is added when the component is loaded.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
         A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>TranscriptionUnit</code> object; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
         the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">EXTENT-UNKNOWN?</th>
        <td align="left"> Ignored.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>TranscriptionUnit</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>TranscriptionUnit</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>

No additional table rows are added based on entries in <KBD>transunits.dat</KBD>.


<a name="Genes"></a>
  <h2><a name="Genes">Genes</a></h2>
  <a name="Genes">Genes are input from the file <kbd>genes.dat</kbd>.
A row is added to the <code>Gene</code> table for each entry in <kbd>genes.dat</kbd>.
  </a>
  <p><a name="Genes"><br>
  </a><a name="genes.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>genes.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code> associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Gene</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Gene</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Gene.Name</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENT-OF[*]</th>
        <td align="left"> If value matches the <b>UNIQUE_ID</b> of a previously loaded transcription unit,
        a row is added to <code>TranscriptionUnitComponent</code>: <BR>
        <CODE>TranscriptionUnitComponent.TranscriptionUnitWID</CODE> is the WID of the <code>TranscriptionUnit</code> object; <BR>
        <CODE>TranscriptionUnitComponent.OtherWID</CODE> is the WID of this <code>Gene</code> object; <br>
        <CODE>TranscriptionUnitComponent.Type</CODE> is 'gene'.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
	A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Gene</code> object; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
	the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">INTERRUPTED</th>
        <td align="left"> <code>Gene.INTERRUPTED</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">LEFT-END-POSITION</th>
        <td align="left"> <code>Gene.CodingRegionStart</code> or <br>
        <code>Gene.CodingRegionEnd</code>, depending on <b>TRANSCRIPTION-DIRECTION</b></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PRODUCT[*]</th>
        <td align="left"> Value should match the <b>UNIQUE_ID</b> of a protein.<br>
        If so, a row is added to <code>GeneWIDProteinWID</code>. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">RIGHT-END-POSITION</th>
        <td align="left"> <code>Gene.CodingRegionEnd</code> or <br>
        <code>Gene.CodingRegionStart</code>, depending on <b>TRANSCRIPTION-DIRECTION</b></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PI</th>
        <td align="left"> <code>Gene.PICalc</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Gene</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
      </tr>
      <tr>
        <th align="left">TRANSCRIPTION-DIRECTION</th>
        <td align="left">  A value of "+" indicates that <b>LEFT-END-POSITION</b> is stored as <br>
        <code>Gene.CodingRegionStart</code> and that <b>RIGHT-END-POSITION</b> is stored as <br>
        <code>Gene.CodingRegionEnd</code>.<br>
	A value of "-" indicates that <b>LEFT-END-POSITION</b> is stored as <br>
        <code>Gene.CodingRegionEnd</code> and that <b>RIGHT-END-POSITION</b> is stored as <br>
        <code>Gene.CodingRegionStart</code>.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">TYPES[*]</th>
        <td align="left"> Ignored unless the <A HREF="../../../metacyc-ontology-loader/doc/metacyc-ontology-manual.html">MetaCyc Ontology</A> dataset has been previously loaded; each type is a <CODE>Term.Name</CODE> from the Multifun subontology of that dataset.
	A row is added to <code>RelatedTerm</code>: <br>
        <code>RelatedTerm.TermWID</code> references a term that is this type. <br>
        <code>RelatedTerm.OtherWID</code> is the WID of this <code>Gene</code> object. <BR>
        <code>RelatedTerm.Relationship</code> is 'superclass'</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>Gene.GenomeID</code> and <br>
        <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Gene</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>

  <h3>Additional Tables</h3>
A row is added to <code>GeneWIDProteinWID</code> for each <b>PRODUCT</b>
attribute, associating the gene with the gene product.
  <p>A row is added to <code>BioSourceWIDGeneWID</code> for each gene
entry. <code>BioSourceWIDGeneWID.BioSourceWID</code> is the WID from
the one row of <code>BioSource</code> created by the loader for the
species being loaded. 
 <P>
A row is added to <CODE>TranscriptionUnitComponent</CODE> when a <B>COMPONENT-OF</B> attribute matches the
<B>UNIQUE-ID</B> of a previously loaded transcription unit.
 <P>
If the MetaCyc Ontology dataset has been loaded,
a row is added to <CODE>RelatedTerm</CODE> for each <B>TYPES</B> attribute that is a term
in the Multifun ontology that is part of the MetaCyc Ontology.


<h2><a name="Promoters">Promoters</a></h2>
Promoters are input from the file <kbd>promoters.dat</kbd>.
A row is added to the <code>Feature</code> table for each entry in this file:
<UL>
<LI> <code>Feature.Type</code> NULL. 
<LI> <code>Feature.Class</code> is 'promoter'. 
<LI> <code>Feature.SequenceType</code> is 'N'. 
<LI> <code>Feature.SequenceWID</code> is NULL. 
<LI> <code>Feature.RegionOrPoint</code> is 'point'. 
<LI> <code>Feature.PointType</code> is 'center'. 
</UL>

  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>promoters.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
      <tr>
        <th align="left">ABSOLUTE-PLUS-1-POSITION</th>
        <td align="left"><code>Feature.StartPosition</code> and <code>Feature.EndPosition</code>.<BR>
	This is the position at which transcription starts.</td>
        <td align="left"><br>
        </td>
      </tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Feature</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Feature.Description</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENT-OF[*]</th>
        <td align="left"> If value matches the <b>UNIQUE_ID</b> of a previously loaded transcription unit,
        a row is added to <code>TranscriptionUnitComponent</code>: <BR>
        <CODE>TranscriptionUnitComponent.TranscriptionUnitWID</CODE> is the WID of the <code>TranscriptionUnit</code> object; <BR>
        <CODE>TranscriptionUnitComponent.OtherWID</CODE> is the WID of this <code>Feature</code> object; <br>
        <CODE>TranscriptionUnitComponent.Type</CODE> is 'promoter'.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements. A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Feature</code> object; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PROMOTER-EVIDENCE</th>
        <td align="left"> Ignored. There is typically an associated <B>CITATIONS</B> attribute for thie evidence code. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REGULATED-BY</th>
        <td align="left"> Ignored. This is the converse of the <B>REGULATED-ENTITY</B> attribute for a transcriptional <A HREF="#Regulation">regulator</A>. </td>
        <td align="left"><br>
        </td>
      </tr>
       <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>

A row is added to <CODE>TranscriptionUnitComponent</CODE> for each <B>COMPONENT-OF</B> attribute
that references a previously loaded transcription unit.


<h2><a name="Terminators">Terminators</a></h2>
Terminators are input from the file <kbd>terminators.dat</kbd>.
A row is added to the <code>Feature</code> table for each entry in this file:
<UL>
<LI> <code>Feature.Type</code> NULL. 
<LI> <code>Feature.Class</code> is 'terminator'. 
<LI> <code>Feature.SequenceType</code> is 'N'. 
<LI> <code>Feature.SequenceWID</code> is NULL. 
<LI> <code>Feature.RegionOrPoint</code> is 'region'. 
</UL>

  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>terminators.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Feature</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Feature.Description</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENT-OF[*]</th>
        <td align="left"> If value matches the <b>UNIQUE_ID</b> of a previously loaded transcription unit,
        a row is added to <code>TranscriptionUnitComponent</code>: <BR>
        <CODE>TranscriptionUnitComponent.TranscriptionUnitWID</CODE> is the WID of the <code>TranscriptionUnit</code> object; <BR>
        <CODE>TranscriptionUnitComponent.OtherWID</CODE> is the WID of this <code>Feature</code> object; <br>
        <CODE>TranscriptionUnitComponent.Type</CODE> is 'terminator'.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements. A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Feature</code> object; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">LEFT-END-POSITION</th>
        <td align="left"> <code>Feature.StartPosition</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">RIGHT-END-POSITION</th>
        <td align="left"> <code>Feature.EndPosition</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>

A row is added to <CODE>TranscriptionUnitComponent</CODE> for each <B>COMPONENT-OF</B> attribute
that references a previously loaded transcription unit.

<h2><a name="DNABindsites">DNA Binding Sites</a></h2>
DNA binding sites are input from the file <kbd>dnabindsites.dat</kbd>.
A row is added to the <code>Feature</code> table for each entry in this file:
<P>
The TRANSCRIPTION-DIRECTION attribute is ignored for transcription units.
In particular, neither it nor the TRANSCRIPTION-DIRECTION of the associated gene is
used in computing Feature.StartPosition of Feature.EndPosition for DNA binding sites.
This could lead to incorrect values for these columns.

<UL>
<LI> <code>Feature.Type</code> is NULL. 
<LI> <code>Feature.Class</code> is always 'binding site'. 
<LI> <code>Feature.SequenceType</code> is 'N'. 
<LI> <code>Feature.SequenceWID</code> is NULL. 
<LI> <code>Feature.RegionOrPoint</code> is 'point'. 
<LI> <code>Feature.PointType</code> is 'center'. 
</UL>
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>dnabindsites.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Feature</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Feature.Description</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMPONENT-OF[*]</th>
        <td align="left"> If value matches the <b>UNIQUE_ID</b> of a previously loaded transcription unit,
        a row is added to <code>TranscriptionUnitComponent</code>: <BR>
        <CODE>TranscriptionUnitComponent.TranscriptionUnitWID</CODE> is the WID of the <code>TranscriptionUnit</code> object; <BR>
        <CODE>TranscriptionUnitComponent.OtherWID</CODE> is the WID of this <code>Feature</code> object; <br>
        <CODE>TranscriptionUnitComponent.Type</CODE> is 'binding site'.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements. A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Feature</code> object; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REGULATED-PROMOTER</th>
        <td align="left">References the <B>UNIQUE-ID</B> of a promoter. Assuming that promoter has been
	previously loaded, its <B>ABSOLUTE-PLUS-1-POSITION</B> is used to convert this binding site's
	<B>RELATIVE-CENTER-POSITION</B> from a relative to an absolute position.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">RELATIVE-CENTER-POSITION</th>
        <td align="left"> This numeric value designates either an integral position, or a position halfway between
	two integral positions. First of all, if the attribute is positive, one is subtracted from it.
	It is then added to the  <code>Feature.StartPosition</code> of the
	promoter named in the <B>REGULATED-PROMOTER</B> attribute. If the sum is integral it is stored in both
	<code>Feature.StartPosition</code> and <code>Feature.EndPosition</code>. If it is nonintegral,
	the next-lowest integer is stored in <code>Feature.StartPosition</code> and
	the next-highest integer is stored in <code>Feature.EndPosition</code>. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Feature</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Feature</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>

A row is added to <CODE>TranscriptionUnitComponent</CODE> for each <B>COMPONENT-OF</B> attribute
that references a previously loaded transcription unit.


 <a name="Reactions"></a>
<h2><a name="Reactions">Reactions</a></h2>
Reactions are input from the file <kbd>reactions.dat</kbd>.
A row is added to the <code>Reaction</code> table for each entry in <kbd>reactions.dat</kbd>.
  </a>
  <p><a name="Reactions"><br>
  </a><a name="reactions.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>reactions.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">BALANCE-STATE</th>
        <td align="left"> Ignored
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Reaction</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">^COEFFICIENT[*]</th>
        <td align="left"> <code>Reactant.Coefficient</code> or <code>Product.Coefficient</code>
        <br>
for the immediately preceding <b>LEFT</b> or <b>RIGHT</b> attribute. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Reaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Reaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Reaction</code>
object; <br>
        <code>CrossReference.DatabaseName</code> is the first element
of the list; <br>
        <code>CrossReference.XID</code> is the second element of the
list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DELTAG0</th>
        <td align="left"> <code>Reaction.DeltaG</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">EC-NUMBER</th>
        <td align="left"> <code>Reaction.ECNumber</code> or <code>Reaction.ECNumberProposed</code>,
        <br>
depending on <b>OFFICIAL-EC?</b></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">OFFICIAL-EC?</th>
        <td align="left"> Determines whether <b>EC-NUMBER</b> is
stored as <br>
        <code>Reaction.ECNumber</code> or <code>Reaction.ECNumberProposed</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">ORPHAN?</th>
        <td align="left"> Ignored
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">LEFT[*]</th>
        <td align="left">
A row is added to <code>Reactant</code>; its <CODE>ReactionWID</CODE> is the WID of this reaction. <br>
The attribute designates a substrate or a class of substrates involved in the reaction.
It is translated as discussed below (see Additional Tables); the WID for the substrate
is stored in  <code>Reactant.OtherWID</code>.
If a <b>^COEFFICIENT</b> attribute follows immediately, it is stored
as <code>Reactant.Coefficient</code>. Otherwise the value 1 is stored.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">RIGHT[*]</th>
        <td align="left">
A row is added to <code>Product</code>; its <CODE>ReactionWID</CODE> is the WID of this reaction.. <br>
The attribute designates a substrate or a class of substrates involved in the reaction.
It is translated as discussed below (see Additional Tables); the WID for the substrate
is stored in  <code>Product.OtherWID</code>.
If a <b>^COEFFICIENT</b> attribute follows immediately, it is stored
as <code>Product.Coefficient</code>. Otherwise the value 1 is stored.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SPONTANEOUS?</th>
        <td align="left"> <code>Reaction.Spontaneous</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Reaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Reaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>
As noted, a row is added to <code>Reactant</code> for each <b>LEFT</b>
attribute and <code>Product</code> for each <b>RIGHT</b> attribute.
If the attribute matches the
<B>UNIQUE-ID</B>, the <B>COMMON-NAME</B>, or a <B>SYNONYM</B>
of a <code>Chemical</code> or <code>Protein</code>, its WID is
stored as <code>Reactant.WID</code>.
Else it is assumed to specify a class of chemicals;
an entry in <CODE>Chemical</CODE> is created for it, such that <CODE>Chemical.Name</CODE>
is the attribute, and <CODE>Chemical.Class</CODE> is 'T'. 

  <h2><a name="EnzRxns">Enzymatic Reactions</a></h2>
  <a name="EnzRxns">Enzymatic reactions are input from the file <kbd>enzrxns.dat</kbd>.
A row is added to the <code>EnzymaticReaction</code> table for each
entry in <kbd>enzrxns.dat</kbd>. </a>
  <p><a name="EnzRxns"><br>
  </a><a name="enzrxns.dat">
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>enzrxns.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">ALTERNATIVE-COFACTORS[*] <br>
ALTERNATIVE-SUBSTRATES[*]</th>
        <td align="left"> The attribute consists of a list of names <b>(PRIMARY
ALT1 ALT2 ... ALTn)</b>. <br>
For each name, the <code>Chemical</code> table is queried for a <b>UNIQUE_ID</b>
or a <b>COMMON_NAME</b>of a compound in the
database being loaded. If none is found, a row is added to <code>Chemical</code>;
the name is stored as <code>Chemical.Name</code>; its <code>Chemical.WID</code>
is used as described below. <br>
        <br>
N rows are added to <code>EnzReactionAltCompound</code>, one for each
        <b>ALTi</b>: <br>
        <code>EnzReactionAltCompound.EnzymaticReactionWID</code> is
the WID of this enzymatic reaction; <br>
        <code>EnzReactionAltCompound.PrimaryWID</code> is the WID
associated with the primary compound; <br>
        <code>EnzReactionAltCompound.AlternativeWID</code> is the WID
associated with compound <b>ALTi</b>; <br>
        <code>EnzReactionAltCompound.Cofactor</code> is 'T' for <b>ALTERNATIVE-COFACTORS</b>,
'F' for <b>ALTERNATIVE-SUBSTRATES</b>.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>EnzymaticReaction</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COFACTORS[*] <br>
PROSTHETIC-GROUPS[*]
COFACTORS-OR-PROSTHETIC-GROUPS[*]</th>
        <td align="left"> The <code>Chemical</code> and <code>Protein</code>
tables (in that order) are queried for a <b>UNIQUE_ID</b> or a <b>COMMON_NAME</b>of
a compound or a
protein in the database being loaded
that matches the value of this attribute. If none is found, a row is
added to <code>Chemical</code>; the attribute value is stored as <code>Chemical.Name</code>
its <code>Chemical.WID</code> is used as described below. <br>
        <br>
A row is added to <code>EnzReactionCofactor</code>; <code>EnzReactionCofactor.EnzymaticReactionWID</code>
is the WID of this enzymatic reaction; <code>EnzReactionCofactor.CompoundWID</code>
is the WID associated with the compound or protein; <code>EnzReactionCofactor.Prosthetic</code>
is 'F' <b>COFACTORS</b>, 'T' for <b>PROSTHETIC-GROUPS</b>, and NULL
for <b>COFACTORS-OR-PROSTHETIC-GROUPS</b>. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>EnzymaticReaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> Ignored</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REQUIRED-PROTEIN-COMPLEX</th>
        <td align="left"> Value should match the <b>UNIQUE_ID</b> of a
protein. <br>
If so, the WID of the protein is stored as <code>EnzymaticReaction.ComplexWID</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>EnzymaticReaction</code>
object; <br>
        <code>CrossReference.DatabaseName</code> is the first element
of the list; <br>
        <code>CrossReference.XID</code> is the second element of the
list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">ENZYME</th>
        <td align="left"> Required. Value should match the <b>UNIQUE_ID</b>
of a protein. <br>
If so, the WID of the protein is stored as <code>EnzymaticReaction.ProteinWID</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REACTION</th>
        <td align="left"> Required. Value should match the <b>UNIQUE_ID</b>
of a reaction. <br>
If so, the WID of the reaction is stored as <code>EnzymaticReaction.ReactionWID</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REACTION-DIRECTION</th>
        <td align="left"> <code>EnzymaticReaction.ReactionDirection</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">REGULATED-BY</th>
        <td align="left"> Ignored. This is the converse of the <B>REGULATED-ENTITY</B> attribute for a enzymatic <A HREF="#Regulation">regulator</A>. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Reaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>EnzymaticReaction</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  <h3>Additional Tables</h3>
The loader adds rows to <code>EnzReactionAltCompound</code>, <code>EnzReactionCofactor</code>,
and <code>EnzReactionInhibitorActivator</code>
as noted above.


<h2><a name="Regulation">Regulation</a></h2>

A regulation entry describes a relationship between a regulated entity and an agent
that performs regulation. There are two types of regulation that are represented in the database
and translated to the BioWarehouse -- regulation of transcription, and regulation of enzyme activity.
The regulation type is indicated by the <B>TYPES</B> attribute.
 <P>
In transcriptional regulation, the regulator is a protein that is a transcription factor;
the regulated entity is a promoter that is a component of one or more transcription units.
In enzymatic regulation, the regulator is a chemical; the regulated entity is an enzymatic reaction.
Comments, citations, synonyms, and crossreferences of each entry are linked to the regulator.
Regulator proteins and compounds will have multiple <CODE>DBID</CODE>s -- their <B>UNIQUE-ID</B>
from this entry as well as the <B>UNIQUE-ID</B> from the <A HREF="#Proteins">protein</A>
or <A HREF="#Compounds">compound</A> entry.
 <P>
All regulation is characterized by a mode, indicating whether the process is inhibited or activated.
In addition, enzymatic regulation is characterized by a regulation mechanism, as well as a flag
indicating physiological relevance.
<P>
Note that for transcriptional regulation, no rows are added to the BioWarehouse. It has no representation
of transcription factors. However, the naming conventions of BioCyc database may be exploited to find all
transcription factors by finding all proteins that have a <CODE>DBID.XID</CODE> starting with 'REG'.
 <P>

  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>regulation.dat</kbd></b> where <B>TYPES</B> is 'Regulation-of-Transcription-Initiation'</caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">ASSOCIATED-BINDING-SITE</th>
        <td align="left">Ignored. </td>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of the
<code>Protein</code> referenced by the <B>REGULATOR</B> attribute.
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of the
<code>Protein</code> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements. A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of the <code>Protein</code> referenced by the <B>REGULATOR</B> attribute; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">MODE</th>
        <td align="left">Ignored. </td>
      </tr>
      <tr>
        <th align="left">REGULATED-ENTITY</th>
        <td align="left">References the <B>UNIQUE-ID</B> of a <a href="#Promoters">promoter</a> <CODE>Feature</CODE>
          that is a component of a <a href="#TransUnits">transcription unit</a>.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
      <tr>
        <th align="left">REGULATOR</th>
        <td align="left">References the <B>UNIQUE-ID</B> of a  <a href="#Proteins">protein</a> that is a transcription factor.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of the <code>Protein</code> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">TYPES</th>
        <td align="left">'Regulation-of-Transcription-Initiation'. Determines whether this entry is translated as
transcriptional or enzymatic regulation.
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of the <code>Protein</code> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>

 <P> <BR>

For each entry that describes enzymatic regulation, a row is added to <code>EnzReactionInhibitorActivator</code>.
Entry attributes determine the column values as described in the table below.

 <P>
  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>regulation.dat</kbd></b> where <B>TYPES</B> is 'Regulation-of-Enzyme-Activity'</caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of the
<code>Chemical</code> or <CODE>Protein</CODE> referenced by the <B>REGULATOR</B> attribute.
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of the
<code>Chemical</code> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements. A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of the <code>Chemical</code> or <CODE>Protein</CODE> referenced by the <B>REGULATOR</B> attribute; <br>
        <code>CrossReference.DatabaseName</code> is the first element of the list; <br>
        <code>CrossReference.XID</code> is the second element of the list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">MECHANISM</th>
        <td align="left">
        <code>EnzReactionInhibitorActivator.Mechanism</code> is
        <ul>
          <li> 'A' if <b>MECHANISM</b> is ':ALLOSTERIC',
          <li> 'C' if <b>MECHANISM</b> is ':COMPETITIVE',
          <li> 'I' if <b>MECHANISM</b> is ':IRREVERSIBLE'
          <li> 'N' if <b>MECHANISM</b> is ':NONCOMPETITIVE' or ':UNCOMPETITIVE',
          <li> 'U' if <b>MECHANISM</b> is ':UNKMECH',
	  <LI> NULL otherwise.	
       </ul>
        </td>
       </tr>
      <tr>
        <th align="left">MODE</th>
        <td align="left">
        <code>EnzReactionInhibitorActivator.InhibitOrActivate</code> is
        <ul>
	<LI> 'A' if <b>MODE</b> is '+', <BR>
	<LI> 'I' if <b>MODE</b> is '-', and <BR>
	<LI> NULL otherwise.
	</ul>
        </td>
      </tr>
       <tr>
        <th align="left">PHYSIOLOGICALLY-RELEVANT?</th>
        <td align="left">
	<code>EnzReactionInhibitorActivator.PhysioRelevant</code> is
        <ul>
	<LI> 'T' if <b>MODE</b> is 'T', <BR>
	<LI> 'F' otherwise.
	</ul>
 </td>
      </tr>
     <tr>
        <th align="left">REGULATED-ENTITY</th>
        <td align="left">References the <B>UNIQUE-ID</B> of an <a href="#EnzRxns">enzymatic reaction</a>. <BR>
        Its WID is <code>EnzReactionInhibitorActivator.EnzymaticReactionWID</code>.
        </td>
      </tr>
      <tr>
      <tr>
        <th align="left">REGULATOR</th>
        <td align="left"> The <code>Chemical</code> table is queried
	for a <b>UNIQUE_ID</b> or a <b>COMMON_NAME</b> of a compound in the
	database being loaded that matches the value of this attribute. If
	none is found, a row is added to <code>Chemical</code>; the attribute
	value is stored as <code>Chemical.Name</code> its
	<code>Chemical.WID</code> is
        <code>EnzReactionInhibitorActivator.CompoundWID</code><br>
       </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of the <code>Chemical</code> or <CODE>Protein</CODE> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">TYPES</th>
        <td align="left">'Regulation-of-Enzyme-Activity'. Determines whether this entry is translated as
transcriptional or enzymatic regulation.
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of the <code>Chemical</code> or <CODE>Protein</CODE> referenced by the <B>REGULATOR</B> attribute.</td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </a> </p>


<h2><a name="Pathways">Pathways</a></h2>

Pathways are input from the file <kbd>pathways.dat</kbd>.
A row is added to the <code>Pathway</code> table for each entry in <kbd>pathways.dat</kbd>.
The <code>Pathway.Type</code> column value of each row is set to 'O'
to signify the pathway is from a real organism. <code>Pathway.BioSourceWID</code>
is the WID from the one row of <code>BioSource</code> created by the
loader for the species being loaded. </a>
  <p>Pathway entries can reference other pathways
(using their <b>UNIQUE-ID</b>), and there is no guarantee that a
pathway entry will be defined before a reference to it occurs. The
loader adds a row to <code>Pathway</code> and assigns it a WID upon
the first reference to a pathway, and performs a SQL UPDATE of the row
when its entry is fully defined.</p>

  <table border="5">
    <caption align="top"><b>Translation semantics for <kbd>pathways.dat</kbd></b></caption>
    <tbody>
      <tr>
        <th align="left">BioCyc Attribute</th>
        <th align="left">Warehouse Semantics</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <th align="left">CITATIONS[*]</th>
        <td align="left"> Each attribute is either an evidence code or
the <b>UNIQUE-ID</b> of a publication.
See <a href="#SupportTable">Support Table</a> for translation of evidence codes.
Each publication <b>UNIQUE-ID</b>
is possibly enclosed in square brackets,
and possibly missing the leading "PUB-". <br>
A row is added to <code>CitationWIDOtherWID</code>; <br>
        <code>CitationWIDOtherWID.CitationWID</code> is the WID of the <code>Citation</code>
associated with this unique ID; <br>
        <code>CitationWIDOtherWID.OtherWID</code> is the WID of this <code>Pathway</code>
        <br>
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMENT[*]</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Pathway</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">COMMON-NAME</th>
        <td align="left"> <code>Pathway.Name</code></td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">DBLINKS[*]</th>
        <td align="left"> Attribute is a list for two or more elements.
A row is added to <code>CrossReference</code>; <br>
        <code>CrossReference.OtherWID</code> is the WID of this <code>Pathway</code>
object; <br>
        <code>CrossReference.DatabaseName</code> is the first element
of the list; <br>
        <code>CrossReference.XID</code> is the second element of the
list; <br>
the rest of the list is ignored. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">HYPOTHETICAL-REACTIONS[*]</th>
        <td align="left"> Value should match the <b>UNIQUE_ID</b> of a reaction. <br>
If so, <code>PathwayReaction.Hypothetical</code> is 'T' <br>
for the row added to <code>PathwayReaction</code> for the reaction.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">NET-REACTION-EQUATION</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Pathway</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PATHWAY-INTERACTIONS</th>
        <td align="left"> <code>CommentTable.Comm</code>; <br>
        <code>CommentTable.OtherWID</code> is the WID of this <code>Pathway</code> object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PATHWAY-LINKS[*]</th>
        <td align="left"> Indicates pathways that are linked via a common substrate. <br>
        If value is of form (Unique-ID) it is probably a pathway reference and is ignored. Otherwise 
	value is of form (Compound PathwaySpec1 ... PathwaySpecN) where each PathwaySpec is either
        a descriptor or (descriptor . direction). The direction is ignored. The descriptor may be either a quoted
        string or a <B>UNIQUE-ID</B> of some BioCyc object (not necessarily a pathway).
	If the descriptor is anything other than a <B>UNIQUE-ID</B> for a previously defined pathway,
        a row is added to <code>Pathway</code>: <br>
        <code>Pathway.Name</code> is the descriptor. <br>
	A row is added to <code>PathwayLink</code> for each PathwaySpec: <br>
        <code>PathwayLink.ChemicalWID</code> is the WID for the Compound; <br>
        <code>PathwayLink.Pathway1WID</code> is the WID of this <code>Pathway</code> object; <br>
        <code>PathwayLink.Pathway2WID</code> is the WID of the linked <code>Pathway</code>. <br>
        During postprocessing, any <CODE>Pathway</CODE> rows that were added that were not actually pathways
        (i.e., no pathway entry was later encountered for that descriptor) are deleted from <CODE>Pathway</CODE>,
        along with any linked <CODE>PathwayLink</CODE> rows.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">PREDECESSORS[*]</th>
        <td align="left"> Collectively, these specify the graph of reactions that form
	the pathway. Each value is of one of two forms: <BR>
	<OL>
	<LI> (Successor Predecessor1 ... PredecessorN); zero predecessors are allowed
	<LI> Predecessor-PathwayID
	</OL>
	For case 1: <BR>
	A row is added to <code>PathwayReaction</code> for each Predecessor.
	For each row: <br>
        <code>PathwayReaction.PathwayWID</code> is the WID of this <code>Pathway</code> object; <br>
        <code>PathwayReaction.ReactionWID</code> is the <code>Reaction</code> WID of Successor; <br>
        <code>PathwayReaction.Hypothetical</code> is 'F', unless successor is named as a <b>HYPOTHETICAL-REACTIONS</b> attribute; <br>
        <code>PathwayReaction.PriorReactionWID</code> is the <code>Reaction</code> WID of Predecessor, or NULL if there are none. <BR>
	For case 2: <BR>
	Each such attribute should also occur as a <B>SUB-PATHWAYS</B> attribute. If it does not, it is ignored.
	</td>
      </tr>
      <tr>
	<th align="left">REACTION-LIST[*]</th>
	<td align="left"> Each attribute is a <B>UNIQUE-ID</B> of a reaction or pathway.
	Pathways occurring here are ignored. 
	For each reaction occurring here, but not occurring as a Successor in
	a <B>PREDECESSORS</B> attribute, a row is added to <CODE>PathwayReaction</CODE>: <BR>
	<code>PathwayReaction.PathwayWID</code> is the WID of this <code>Pathway</code> object; <br>
	<code>PathwayReaction.ReactionWID</code> is the <code>Reaction</code>
	WID of the attribute; <br>
	<code>PathwayReaction.Hypothetical</code> is 'F', unless
	the attribute is named as a <b>HYPOTHETICAL-REACTIONS</b> attribute; <br>
	<code>PathwayReaction.PriorReactionWID</code> is NULL.
	
	</td>
	</tr>
	<tr>
	<th align="left">SUB-PATHWAYS[*]</th>
	<td align="left"> 
	This pathway inherits the reaction graph of the pathway whose
	<B>UNIQUE-ID</B> equals the attribute.
	That is, a <code>PathwayReaction</code> row is added for this pathway for each
	<code>PathwayReaction</code> row of the pathway designated by the attribute.
	The columns of each <CODE>PathwayReaction</CODE> row are identical,
	except that <CODE>PathwayWID</CODE> is changed from the
	attribute's pathway WID to the WID of this pathway. <BR>
	Note: sub/superpathway information is loaded via <B>SUPER-PATHWAYS</B>.
	</td>
	</tr>
      <tr>
        <th align="left">SUPER-PATHWAYS[*]</th>
        <td align="left"> Value should match the <b>UNIQUE_ID</b> of a pathway. <br>
If so, a row is added to <code>SuperPathway</code>: <br>
        <code>SuperPathway.SuperPathwayWID</code> is the WID associated with this <b>UNIQUE_ID</b>; <br>
        <code>SuperPathway.PathwayWID</code> is the WID of this <code>Pathway</code> object. </td>
      </tr>
      <tr>
        <th align="left">SYNONYMS[*]</th>
        <td align="left"> <code>SynonymTable.Syn</code>; <br>
        <code>SynonymTable.OtherWID</code> is the WID of this <code>Pathway</code>
object</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <th align="left">UNIQUE-ID</th>
        <td align="left"> <code>DBID.XID</code>; <br>
        <code>DBID.OtherWID</code> is the WID of this <code>Pathway</code>
object</td>
      </tr>
    </tbody>
  </table>
  </a> </p>
  
  <h3>Additional Tables</h3>
Rows are added to the tables <code>SuperPathway</code>, <code>PathwayReaction</code>,
and <code>PathwayLink</code> as specified above.


  <h2><a name="SupportTable">Support Table</a></h2>
Most BioCyc entries allow a <B>CITATIONS</B> attribute to be specified.
If this attribute begins with a colon or contains the string ':EV-' anywhere,
it is treated as an indicator of evidence
for the validity of the associated entry, or some portion of it.
The evidence code is the text between the colon and the next colon (if present) or
the end of the attribute (excluding the colons).
<p>
For each attribute of this form, a row in the <code>Support</code> table is created as follows:
  <p>
  <table border="5">
    <caption align="top"><b>Column values for <code>Support</code> row</b></caption>
    <tbody>
      <tr>
        <th align="left">Column</th>
        <th align="left">Value assigned by BioCyc loader</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <td align="left"><code>WID</code> </td>
        <td align="left"> The BioWarehouse ID allocated for this Support.</td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>OtherWID</code> </td>
        <td align="left"> The WID of the entry this supporting evidence applies to. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Type</code> </td>
        <td align="left"> The evidence code (e.g., 'EV-EXP-IMP-POLAR-MUTATION').<BR>
	<I>Note: this is not consistent with the schema documentation, which states this column is either 'computational or 'experimental'.</I> </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>Confidence</code> </td>
        <td align="left"> NULL </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>DatasetWID</code> </td>
        <td align="left"> The value <code>Dataset.WID</code> assigned to the dataset being loaded. </td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </p>
  <p> </p>

  <h3>Additional Tables</h3>

If the <B>CITATIONS</B> attribute contains a reference to a publication ID,
a row is added to the <code>CitationWIDOtherWID</code> to associate the <code>Support</code> row
with the <code>Citation</code> of the publication.


  <h2><a name="EntryTable">Entry Table</a></h2>
For each object loaded from the database, a row
in the <code>Entry</code> table is created as follows:
  <p>
  <table border="5">
    <caption align="top"><b>Column values for <code>Entry</code> row</b></caption>
    <tbody>
      <tr>
        <th align="left">Column</th>
        <th align="left">Value assigned by BioCyc loader</th>
        <th align="left"><br>
        </th>
      </tr>
      <tr>
        <td align="left"><code>OtherWID</code> </td>
        <td align="left"> The WID of the entry described by this row.
Entry may be in <br>
        <code>Chemical</code>, <code>Reaction</code>, <code>Protein</code>,
        <code>Gene</code>, <code>EnzymaticReaction</code>, or <code>Pathway</code>.
        </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>InsertDate</code> </td>
        <td align="left"> The time/date the loader was run. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>CreationDate</code> </td>
        <td align="left"> NULL </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>ModifiedDate</code> </td>
        <td align="left"> NULL </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>LineNumber</code> </td>
        <td align="left"> The line number from the input file on which this entry began. </td>
        <td align="left"><br>
        </td>
      </tr>
       <tr>
        <td align="left"><code>LoadError</code> </td>
        <td align="left"> 'T' if a parse error is detected, 'F'
otherwise. </td>
        <td align="left"><br>
        </td>
      </tr>
      <tr>
        <td align="left"><code>DatasetWID</code> </td>
        <td align="left"> The value <code>Dataset.WID</code> assigned
to the dataset being loaded. </td>
        <td align="left"><br>
        </td>
      </tr>
    </tbody>
  </table>
  </p>
  <p> </p>

  <hr><a name="Reference"> </a>
  <h1><a name="Reference">References</a></h1>
  <a name="Reference"> </a></li>
<li><a href="https://community.biospice.org/">BioSPICE Web
Site</a> </li>
<li><a href="http://biocyc.org/">BioCyc.org</a> </li>
<li><a href="http://bioinformatics.ai.sri.com/ptools/">Pathway/Genome
Databases (PGDBs)</a> </li>
<li><a href="../../../doc/Schema.html">Schema
documentation</a></li>
</body>
</html>
